### PR TITLE
[Schema 3] Adders and Unit Tests

### PR DESCRIPTION
**Pull request by afein**
Part 3 of #342 
Includes the following:
- Add methods for Namespaces, Nodes and Pods
- Unit tests for adders and the existing util functions

Changes based on comments from #358:
- lock moved inside the cluster object again
- Get methods have more verbose comments, clearly explaining inputs and outputs

@vishh @rjnagal the next PR, Schema 4, will include utility functions that operate on TimeStores, along with their unit tests.

### CODE REVIEW

File: schema/util.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ import (
 	"github.com/GoogleCloudPlatform/heapster/store"
 )
 
+// maxTimestamp returns its largest time.Time argument
```
Comment (vishh) [inline schema/util.go]: latest instead of max may be?

File: schema/impl.go [hunk 2]
```diff
@@ -41,19 +38,23 @@ func newRealCluster() *realCluster {
 	return cluster
 }
 
+// GetAllClusterData returns a pointer to the ClusterInfo, along with all of its metrics.
+// GetAllClusterData also returns the latest cluster timestamp, for reuse in GetNew* methods.
 func (rc *realCluster) GetAllClusterData() (*ClusterInfo, time.Time, error) {
```
Comment (vishh) [inline schema/impl.go]: Change to GetClusterData(start, end time.Time) instead?

File: schema/impl.go [hunk 3]
```diff
@@ -91,3 +95,88 @@ func (rc *realCluster) Update(c *cache.Cache) error {
...
+	}
+
+	if _, ok := node.Pods[pod_name]; ok {
+		in_node = true
+	}
+
+	if in_ns && in_node {
+		// Pod already in Namespace and Node maps, return pointer
+		pod_ptr, _ = node.Pods[pod_name]
+	} else if !in_ns && !in_node {
```
Comment (vishh) [inline schema/impl.go]: Can we remove the 'if' condition here?

**Comment by vishh** [comment]
Just a few comments. Overall LGTM

File: schema/impl.go (hunk 2 above)
Comment (afein) [inline schema/impl.go]: Interface-level methods will be redesigned alongside the REST endpoints.
The existing Getters serve mostly as proof-of-concept endpoints. After the redesign, these methods will probably be renamed to internal methods used only for testing.
Estimate: [Schema 6]

Fundamental ideas:
- Different methods for historical and derived data
- (start, end) parameters for historical data
- Getters return (time, error) and accept a callback function which performs the appropriate conversion and copy of the internal types to external types.

File: schema/util.go (hunk 1 above)
Comment (afein) [inline schema/util.go]: done

File: schema/impl.go (hunk 3 above)
Comment (afein) [inline schema/impl.go]: done

File: schema/impl.go (hunk 2 above)
Comment (vishh) [inline schema/impl.go]: Ack.

**Comment by vishh** [comment]
LGTM.

**Pull request merged**